### PR TITLE
Discovery L072CZ-LRWAN1 support

### PR DESCRIPTION
**Pull request by knielsen**
This pull request adds support for the Discovery L072CZ-LRWAN1 board:

  http://www.st.com/en/evaluation-tools/b-l072z-lrwan1.html

### CODE REVIEW

**Comment by fpistm** [comment]
Thanks @knielsen 
I will have a look soon.

**Comment by LMESTM** [comment]
@knielsen Looks great thanks for the PR ! Any plan on your side to add a LORA library for this board ?

**Comment by knielsen** [comment]
Sorry, no plans for a LoRa library from me currently...

**Comment by fpistm** [comment]
Thanks @knielsen for this variant.
I've made some tests and it's ok.
I've only made some cosmetic update in #156.
LD2 is set to red in the schematics but well set to green in the 'Table 4. Assignment of the control ports'
I've added DIS_L072Z as a possible node value.

File: boards.txt [hunk 1]
```diff
@@ -344,6 +344,21 @@ Disco.menu.pnum.DISCO_L475VG_IOT.build.product_line=STM32L475xx
 Disco.menu.pnum.DISCO_L475VG_IOT.build.variant=DISCO_L475VG_IOT
 Disco.menu.pnum.DISCO_L475VG_IOT.build.cmsis_lib_gcc=arm_cortexM4l_math
 
+# DISCO-L072CZ-LRWAN1 board
+# Support: Serial1 (USART1 on PA10, PA9) and Serial=Serial2 (connected to ST-LINK)
+Disco.menu.pnum.DISCO_L072CZ_LRWAN1=Discovery L072CZ-LRWAN1
+Disco.menu.pnum.DISCO_L072CZ_LRWAN1.node=NODE_L072CZ
```
Comment (fpistm) [inline boards.txt]: I've added DIS_L072Z as a possible node value.

File: variants/DISCO_L072CZ_LRWAN1/PeripheralPins.c [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+//  {PA_7,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM3, 2, 0)},  // TIM3_CH2 - Not available on board
+//  {PA_15, TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF5_TIM2, 1, 0)},  // TIM2_CH1 - Not available on board
+//  {PB_0,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM3, 3, 0)},  // TIM3_CH3 - Not available on board
+//  {PB_1,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM3, 4, 0)},  // TIM3_CH4 - Not available on board
+//  {PB_3,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM2, 2, 0)},  // TIM2_CH2 - Not available on board
+//  {PB_4,  TIM22,  STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF4_TIM22, 1, 0)},  // TIM22_CH1 - Not available on board
+//  {PB_4,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM3, 1, 0)},  // TIM3_CH1 - Not available on board
+//    {PB_5,  TIM22,  STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF4_TIM22, 2, 0)},  // TIM22_CH2
+    {PB_5,  TIM3,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF4_TIM3, 2, 0)},  // TIM3_CH2
```
Comment (fpistm) [inline variants/DISCO_L072CZ_LRWAN1/PeripheralPins.c]: Seems TIM3 on this pin do not work. I've tried TIM22 it is ok.
I've debugged but see nothing wrong. I've tried TIM3 with the Servo and it is ok.
I saw in errata sheet an issue with TIM3_ER on other pin but not on this one.

File: variants/DISCO_L072CZ_LRWAN1/readme.txt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+solder bridges SB7 and SB8.
+
+A4 and A5 can be connected to PB9 and PB8 using solder bridges SB11
+and SB12 (note that PB9 and PB8 do not have any analog/ADC
+capabilities though).
+
+An third analog/ADC input is available as A6 on the PA5 pin.
+
+SPI MOSI and MISO are available on digital pins D11 and D12, but the
+SCK pin is D3 (unlike Arduino Uno where it is on D13).
```
Comment (fpistm) [inline variants/DISCO_L072CZ_LRWAN1/readme.txt]: By default SB2 is closed so D13 is connected to PB13 like D3. So SCK is also on D13.
I've tested SPI and it is ok with D13

File: variants/DISCO_L072CZ_LRWAN1/variant.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  PB_5,  //D4
+  PB_7,  //D5
+  PB_2,  //D6
+  PA_8,  //D7
+  PA_9,  //D8
+  PB_12, //D9
+  PB_6,  //D10
+  PB_15, //D11
+  PB_14, //D12
+  PA_5,  //D13
```
Comment (fpistm) [inline variants/DISCO_L072CZ_LRWAN1/variant.cpp]: D13 is PB13

File: variants/DISCO_L072CZ_LRWAN1/variant.h [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+  PB5,  //D4
+  PB7,  //D5
+  PB2,  //D6
+  PA8,  //D7
+  PA9,  //D8
+  PB12, //D9
+  PB6,  //D10
+  PB15, //D11
+  PB14, //D12
+  PA5,  //D13
```
Comment (fpistm) [inline variants/DISCO_L072CZ_LRWAN1/variant.h]: D13 is PB13

File: variants/DISCO_L072CZ_LRWAN1/variant.h [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+
+//SPI definitions
+#define SS                      10
+#define SS1                     4
+#define SS2                     7
+#define SS3                     8
+#define MOSI                    11
+#define MISO                    12
+#define SCK                     3
```
Comment (fpistm) [inline variants/DISCO_L072CZ_LRWAN1/variant.h]: As D13 is equal to D3 I've revert to 13

File: variants/DISCO_L072CZ_LRWAN1/variant.h [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+//I2C Definitions
+#define SDA                     14
+#define SCL                     15
+
+//Timer Definitions
+//Do not use timer used by PWM pins when possible. See PinMap_PWM.
+#define TIMER_TONE              TIM22
+#define TIMER_UART_EMULATED     TIM21
```
Comment (fpistm) [inline variants/DISCO_L072CZ_LRWAN1/variant.h]: I removed the  TIMER_UART_EMULATED  as it need to be reviewed. (move as SoftwareSerial library)

**Comment by LMESTM** [comment]
@knielsen 

> Sorry, no plans for a LoRa library from me currently...

No problem, this is good to know anyway to avoid any parallel developments - thanks again for your contribution.

File: boards.txt (hunk 1 above)
Comment (knielsen) [reply boards.txt]: ok

File: variants/DISCO_L072CZ_LRWAN1/PeripheralPins.c (hunk 2 above)
Comment (knielsen) [reply variants/DISCO_L072CZ_LRWAN1/PeripheralPins.c]: Interesting, thanks for testing this.
TIM22 is TIMER_TONE in variant.h, but not sure that will necessarily be a problem.

File: variants/DISCO_L072CZ_LRWAN1/readme.txt (hunk 3 above)
Comment (knielsen) [reply variants/DISCO_L072CZ_LRWAN1/readme.txt]: Ok, good that you found this from test. The user manual said SB2 is OFF by default.

Comment (fpistm) [reply variants/DISCO_L072CZ_LRWAN1/readme.txt]: Yes user manual has some errors. I will raise issue internally to correct those one

**Comment by knielsen** [comment]
Looks great @fpistm , thanks!

File: variants/DISCO_L072CZ_LRWAN1/PeripheralPins.c (hunk 2 above)
Comment (fpistm) [reply variants/DISCO_L072CZ_LRWAN1/PeripheralPins.c]: Not a pb. 😉

**Comment by fpistm** [comment]
This PR has been merged thanks #156

**Pull request closed without merging**